### ISSUE TITLE
Further restricting a query already limited to a single row should not require ALLOW FILTERING

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#7964]
In a table with `p`, `c` and `r` being a partition key, clustering key and regular column respectively, a query with `WHERE p=1 AND c=1`, which returns a single row (or none), does not require "ALLOW FILTERING". This is correct.

Now, adding to this restriction yet another restriction on the regular column - `WHERE p=1 AND c=1 AND r=1` - should not require `ALLOW FILTERING` either. This query still processes just a single row and either returns it or not.

Currently *both* Scylla and Cassandra require `ALLOW FILTERING` in this second case, and I think it is a bug in both.
It's a very low priority bug both because this is not a very useful query and because it can be circumvented (with no performance penalty) by adding `ALLOW FILTERING`.

The following cql-pytest test in `test_allow_filtering.py` reproduces this issue:

```python
def test_allow_filtering_single_row(cql, table1):
    check_af_optional(cql, table1, 'k=1 AND c=2', lambda row: row.k==1 and row.c==2)
    # The following currently fails, because ALLOW FILTERING is considered mandatory, not optional
    check_af_optional(cql, table1, 'k=1 AND c=2 AND v=2', lambda row: row.k==1 and row.c==2 and row.v==2)
```

**Comment by psarna** [comment]
Well, Scylla is technically correct - you still fetch this row unconditionally and then filter whether to return it to the client or not. It's definitely not a performance-heavy task, but ALLOW FILTERING mentions that the performance *might* be unpredictable, not that it will be. In fact, filtering can often be a much better choice than creating an index or a view , especially if the number of dropped rows is small or the filtered set is really large (because a scan + filtering will likely perform better than a huge number of random reads). All of these optimistic cases still require ALLOW FILTERING, even if Scylla could deduce that the performance penalty would be negligible.

I think that it's simply not worth it to complicate the code with one more corner case. It's also not clear where should we start requiring ALLOW FILTERING. E.g. `WHERE p = 1 AND c > 0 AND c < 42` is also guaranteed to have a small, constant upper bound of rows to filter.

**Comment by nyh** [comment]
Note that "WHERE p=1 AND c=1" _already_ may return either one or zero rows. To return zero rows always requires non-zero work, and that is not considered filtering. In my opinion (and we have other issues about needing to define this better), "filtering" is when returning O(1) results (0 results included...) may require O(N) work. And this doesn't happen here. The performance of this query _is_ predictable - it always takes the amount of time it takes to read one row. Never more.

The second example is an interesting one. `WHERE p=1 AND c>0 AND c<42` is a partition slice, and does NOT require `ALLOW FILTERING` (this works correctly today) because if you read N bytes from disk you return O(N) bytes in the result. But we can modify this to, `WHERE p=1 AND c>0 AND c<42 AND r=1`. If `c` is an integer, indeed we will be filtering up to 42 rows and possibly returning none, and it is debatable whether "42" which low and predictable should be considered differently from "1". But note that a similar query where `c` isn't an integer, but rather a string or floating point, could have an _unbounded_ number of rows matching, and there the addition of `r=1` definitely does constitute filtering. I agree maybe it's not worth our time to chase all these corner cases. This is why I marked this issue with the "low" tag. However, I believe that with the right definitions of which kind of queries can return "unknown amount of results" it can be easy to consistently classify whether a certain query requires filtering or not, without needing to have a long list of weird corner cases.

**Comment by psarna** [comment]
Yup, I meant a  `r=1` in this example too, otherwise it's just a regular slice query :) My point was simply that it's easy to come up with more cases which return a number of rows bounded by a small constant.

**Comment by slivne** [comment]
Doesn't cassandra require the same - if there is an additional restriction to be applied on the coordinator side - ALLOW FILTERING needs to be added (at least this is my understanding).

**Comment by nyh** [comment]
@slivne yes, as I said in the original report, "Currently both Scylla and Cassandra require ALLOW FILTERING in this second case, and I think it is a bug in both.". I also said I think this is a "very low priority bug". So let's not work on it, but not delete it either. People looking for "ALLOW FILTERING" issues should be able to find this one too.
